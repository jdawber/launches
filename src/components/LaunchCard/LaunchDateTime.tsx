import type { LaunchData } from '../../types';

type LaunchDateTimeProps = Pick<LaunchData, 'launchDate'>;

const locale = 'en-GB';

const LaunchDateTime = ({ launchDate }: LaunchDateTimeProps) => (
  <span>
    {launchDate.toLocaleDateString(locale)} / {launchDate.toLocaleTimeString(locale)}
  </span>
);

export { LaunchDateTime };
