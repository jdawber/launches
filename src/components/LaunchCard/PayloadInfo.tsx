import type { LaunchData } from '../../types';

type PayloadInfoProps = Pick<LaunchData, 'payloads'>;

const PayloadInfo = ({ payloads }: PayloadInfoProps) => (
  <>
    <h3>Payload</h3>
    <ul>
      {payloads.map(({ id, type }) => (
        <li key={id}>{type}</li>
      ))}
    </ul>{' '}
  </>
);

export { PayloadInfo };
