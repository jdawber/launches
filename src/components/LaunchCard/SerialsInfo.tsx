import type { LaunchData } from '../../types';

type SerialsInfoProps = Pick<LaunchData, 'coreSerials'>;

const SerialsInfo = ({ coreSerials }: SerialsInfoProps) => (
  <>
    <h3>Core Serials</h3>
    <ul>
      {coreSerials.map((serial, index) => (
        <li key={index}>{serial}</li>
      ))}
    </ul>
  </>
);

export { SerialsInfo };
