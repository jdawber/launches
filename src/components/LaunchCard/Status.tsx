import type { LaunchData } from '../../types';
import styled from 'styled-components';

const Root = styled.div`
  margin-bottom: 1rem;

  @media screen and (min-width: 800px) {
    width: 100%;
  }
`;

type StatusProps = Pick<LaunchData, 'successful' | 'failureReason'>;

const Status = ({ successful, failureReason }: StatusProps) => <Root>{successful ? '✅ Launch successful' : `❌ ${failureReason ? failureReason : ''}`}</Root>;

export { Status };
