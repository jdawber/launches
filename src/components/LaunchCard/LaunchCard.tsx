import React from 'react';
import Image from 'next/image';
import type { LaunchData } from '../../types';
import styled from 'styled-components';

import { Status } from './Status';
import { SerialsInfo } from './SerialsInfo';
import { PayloadInfo } from './PayloadInfo';
import { LaunchDateTime } from './LaunchDateTime';

const Root = styled.article`
  border: 1px solid #333;
  background: #666;
  border-radius: 1rem;
  margin: 1rem 0;
`;

const Header = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: 1rem;

  h2 {
    margin: 0;
  }
`;

const LaunchDataContainer = styled.div`
  margin: 1rem;

  @media screen and (min-width: 800px) {
    display: flex;
    flex-wrap: wrap;
  }
`;

const InfoContainer = styled.div`
  margin-top: 1rem;

  h3 {
    margin: 0;
  }

  ul {
    margin: 0 0 1rem 0;
  }

  @media screen and (min-width: 800px) {
    margin: 0 0 0 2rem;

    h3 {
      margin: 0;
    }
  }
`;

type LaunchProps = LaunchData;

const LaunchCard = ({ id, missionName, launchDate, missionPatch, coreSerials, payloads, successful, failureReason = null }: LaunchProps) => (
  <Root data-id={id}>
    <Header>
      <h2>{missionName}</h2>
      <LaunchDateTime launchDate={launchDate} />
    </Header>
    <LaunchDataContainer>
      <Status successful={successful} failureReason={failureReason} />
      {missionPatch && (
        <div>
          <Image src={missionPatch} width='150' height='150' alt={`${missionName} mission patch`} />
        </div>
      )}
      <InfoContainer>
        <SerialsInfo coreSerials={coreSerials} />
        <PayloadInfo payloads={payloads} />
      </InfoContainer>
    </LaunchDataContainer>
  </Root>
);

export { LaunchCard };
