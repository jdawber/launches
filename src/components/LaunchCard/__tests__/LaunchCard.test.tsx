import { render, screen } from '@testing-library/react';
import { LaunchCard } from '../LaunchCard';

describe('LaunchCard component', () => {
  it('renders data', () => {
    const { container } = render(
      <LaunchCard
        id={1}
        missionName='mission-name'
        launchDate={new Date('December 17, 2022 01:30:00')}
        missionPatch='https://images2.imgbox.com/40/e3/GypSkayF_o.png'
        coreSerials={[]}
        payloads={[]}
        successful={true}
        failureReason={null}
      />
    );

    expect(screen.getByText('mission-name')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
