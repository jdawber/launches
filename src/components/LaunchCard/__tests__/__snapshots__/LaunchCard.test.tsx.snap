// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchCard component renders data 1`] = `
<div>
  <article
    class="sc-gswNZR cYaHV"
    data-id="1"
  >
    <header
      class="sc-dkrFOg nstuo"
    >
      <h2>
        mission-name
      </h2>
      <span>
        17/12/2022
         / 
        01:30:00
      </span>
    </header>
    <div
      class="sc-hLBbgP fQxBFr"
    >
      <div
        class="sc-bcXHqe keKKPJ"
      >
        âœ… Launch successful
      </div>
      <div>
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27150%27%20height=%27150%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="mission-name mission patch"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
      <div
        class="sc-eDvSVe hDUrPk"
      >
        <h3>
          Core Serials
        </h3>
        <ul />
        <h3>
          Payload
        </h3>
        <ul />
         
      </div>
    </div>
  </article>
</div>
`;
